File: 22.Generate_Parentheses/Generate_Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        helper(n, 0, "", res);
        return res;
    }
    
    void helper(int n, int position, string solution, vector<string> &result){
        if(position == 2*n){
            if(isValid(n, solution)){
                result.push_back(solution);
            }
            return;
        }
        if(isValid(n, solution)){
            helper(n, position+1, solution + '(', result);
            helper(n, position+1, solution + ')', result);
        }
        return;
    }
    
    bool isValid(int n, string s){
        int sum = 0;
        int pos_neg = 0;
        for(auto ch: s){
            if(ch == '('){
                sum += 1;
                pos_neg += 1;
            }
            if(ch == ')')
                pos_neg += -1;
            if(pos_neg < 0)
                return false;
        }
        return sum <= n;
    }
};

File: 22.Generate_Parentheses/Generate_Parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        solution = ''
        result = []
        self.dfs(n, 0, solution, result)
        return result
    
    def dfs(self, n, pos, solution, result):
        if pos == 2*n:
            if self.isValid(n, solution):
                result.append(solution)
            return
        if self.isValid(n, solution):
            self.dfs(n, pos+1, solution + '(', result)
            self.dfs(n, pos+1, solution + ')', result)
        else:
            return
        
    def isValid(self, num, solution):
        left = []
        count = 0
        for i in range(len(solution)):
            if solution[i] == '(':
                left.append('(')
                count += 1
            else:
                if len(left) != 0:
                    left.pop()
                else:
                    return False
                
        return count <= num
        
        
            